src/main/rx/srxl2.c

```diff
@@ -0,0 +1,610 @@
+        }
+    } break;
+    };
+
+    if (writeBufferIdx) {
+        result |= RX_FRAME_PROCESSING_REQUIRED;
+    }
+
+    return result;
+}
+
+static bool srxl2ProcessFrame(const rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    UNUSED(rxRuntimeConfig);
+
+    if (writeBufferIdx == 0) {
+        return true;
+    }
+
+    const uint32_t now = micros();
+
+    if (lastIdleTimestamp > lastReceiveTimestamp) {
+        // time sufficient for at least 2 characters has passed
+        if (now - lastReceiveTimestamp > SRXL2_REPLY_QUIESCENCE) {
+            serialWriteBuf(serialPort, writeBuffer, writeBufferIdx);
+            writeBufferIdx = 0;
+        } else {
+          //            DEBUG("not enough time to send 2 characters passed yet, %d us since last receive, %d required\r\n", now - lastReceiveTimestamp, SRXL2_REPLY_QUIESCENCE);
+        }
+    } else {
+        // DEBUG("still receiving a frame, %d %d\r\n", lastIdleTimestamp, lastReceiveTimestamp);
+    }
+
+    return true;
+}
+
+static uint16_t srxl2ReadRawRC(const rxRuntimeConfig_t *rxRuntimeConfig, uint8_t channelIdx)
+{
+    if (channelIdx >= rxRuntimeConfig->channelCount) {
+        return 0;
+    }
+
+    return 988 + ((rxRuntimeConfig->channelData[channelIdx] >> SRXL2_CHANNEL_SHIFT) >> 1);
+}
+
+void srxl2RxWriteData(const void *data, int len)
+{
+    len = MIN(len, (int)sizeof(writeBuffer));
+    memcpy(writeBuffer, data, len);
+    writeBufferIdx = len;
+}
+
+bool srxl2RxInit(const rxConfig_t *rxConfig, rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    static uint16_t channelData[SRXL2_MAX_CHANNELS];
+    for (size_t i = 0; i < SRXL2_MAX_CHANNELS; ++i) {
+        channelData[i] = SRXL2_CHANNEL_CENTER;
+    }
+
+    unitId = rxConfig->srxl2_unit_id;
+    baudRate = rxConfig->srxl2_baud_rate;
+
+    rxRuntimeConfig->channelData = channelData;
+    rxRuntimeConfig->channelCount = SRXL2_MAX_CHANNELS;
+    rxRuntimeConfig->rxRefreshRate = SRXL2_FRAME_PERIOD_US;
+
+    rxRuntimeConfig->rcReadRawFn = srxl2ReadRawRC;
+    rxRuntimeConfig->rcFrameStatusFn = srxl2FrameStatus;
+    rxRuntimeConfig->rcProcessFrameFn = srxl2ProcessFrame;
+
+    const serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_RX_SERIAL);
+    if (!portConfig) {
+        return false;
+    }
+
+    portOptions_e options = SRXL2_PORT_OPTIONS;
+    if (rxConfig->serialrx_inverted) {
+        options |= SERIAL_INVERTED;
+    }
+    if (rxConfig->halfDuplex) {
+        options |= SERIAL_BIDIR;
+    }
+
+    serialPort = openSerialPort(portConfig->identifier, FUNCTION_RX_SERIAL, srxl2DataReceive,
+        NULL, SRXL2_PORT_BAUDRATE_DEFAULT, SRXL2_PORT_MODE, options);
+
+    if (serialPort) {
+        serialPort->idleCallback = srxl2Idle;
+
+        state = ListenForActivity;
+        timeoutTimestamp = micros() + SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US;
+
+        if (rssiSource == RSSI_SOURCE_NONE) {
+            rssiSource = RSSI_SOURCE_RX_PROTOCOL;
+        }
+    }
+
+    /* handshake protocol
```

Comment:
In terms of maintainability I think it is best if helpful comments such as this one are moved to the top of the file, where other developers will see them before they start reading code.

Reply:
Done.